File: ml/import_kaggle_dataset.py
#!/usr/bin/env python3
import os
import sys
import csv
import json
import argparse
from datetime import datetime

# Define standard headers for our internal system
SCHEMA_FACILITIES = ["id", "name", "type", "latitude", "longitude", "contact"]
SCHEMA_INVENTORY = ["name", "quantity", "unit", "minimumThreshold", "lastUpdated"]
SCHEMA_DISEASES = ["id", "title", "title_lg", "icon", "color", "steps_json", "steps_lg_json"]

# Semantic mappings to auto-detect Kaggle/HDX headers
MAPPINGS = {
    "facilities": {
        "name": ["facility", "clinic", "hospital", "name", "facility_name", "hosp_name", "title"],
        "type": ["type", "level", "tier", "facility_type", "class", "category"],
        "latitude": ["latitude", "lat", "y", "coord_lat", "latitude_deg"],
        "longitude": ["longitude", "long", "lon", "x", "coord_long", "longitude_deg"],
        "contact": ["contact", "phone", "telephone", "mobile", "email", "address"]
    },
    "inventory": {
        "name": ["name", "medicine", "drug", "item", "product", "description"],
        "quantity": ["quantity", "qty", "stock", "count", "amount"],
        "unit": ["unit", "pack", "form", "dosage_form"],
        "minimumThreshold": ["minimumthreshold", "min_stock", "threshold", "min_qty", "alert_level"],
        "lastUpdated": ["lastupdated", "updated", "date", "timestamp"]
    },
    "diseases": {
        "id": ["id", "code", "disease_id"],
        "title": ["title", "name", "disease", "ailment", "condition"],
        "title_lg": ["title_lg", "luganda", "local_name", "title_local"],
        "icon": ["icon", "symbol", "image"],
        "color": ["color", "hex", "theme"],
        "steps_json": ["steps_json", "steps", "treatment", "guidelines", "protocol"],
        "steps_lg_json": ["steps_lg_json", "steps_lg", "treatment_lg", "guidelines_lg"]
    }
}

def clean_value(val):
    return val.strip() if val else ""

def auto_detect_mapping(headers, category):
    detected = {}
    category_mappings = MAPPINGS[category]
    
    for target_col, synonyms in category_mappings.items():
        found = False
        for syn in synonyms:
            for h in headers:
                if syn.lower() == h.strip().lower() or syn.lower() in h.strip().lower().replace(" ", "_"):
                    detected[target_col] = h
                    found = True
                    break
            if found:
                break
    return detected

def prompt_manual_mapping(headers, category, auto_mapped):
    print(f"\n--- Column Mapping Configuration for [{category.upper()}] ---")
    print("Detected columns in raw CSV:")
    for idx, h in enumerate(headers):
        print(f"  [{idx}] {h}")
    print("\nWe need to map these to our database schema. Press Enter to accept auto-detection.")
    
    final_mapping = {}
    target_schema = SCHEMA_FACILITIES if category == "facilities" else (SCHEMA_INVENTORY if category == "inventory" else SCHEMA_DISEASES)
    
    for target in target_schema:
        # Don't strictly require 'id' or fields that can be auto-generated or left blank
        default_val = auto_mapped.get(target, "")
        default_prompt = f" (Auto-detected: '{default_val}')" if default_val else ""
        
        while True:
            choice = input(f"Select source column for '{target}'{default_prompt} [Enter Index or Column Name, or 'skip']: ").strip()
            
            if choice == "":
                if default_val:
                    final_mapping[target] = default_val
                    break
                elif target in ["contact", "title_lg", "steps_lg_json", "id", "icon", "color"]:
                    print(f"Skipping optional field '{target}'")
                    final_mapping[target] = None
                    break
                else:
                    print(f"Error: Field '{target}' is required. Please map it.")
            elif choice.lower() == "skip":
                final_mapping[target] = None
                break
            else:
                # Check if index
                try:
                    idx = int(choice)
                    if 0 <= idx < len(headers):
                        final_mapping[target] = headers[idx]
                        break
                    else:
                        print("Invalid index. Try again.")
                except ValueError:
                    # Check if exact match
                    matching_headers = [h for h in headers if h.lower() == choice.lower()]
                    if matching_headers:
                        final_mapping[target] = matching_headers[0]
                        break
                    else:
                        print(f"Header '{choice}' not found in raw CSV. Try again.")
                        
    return final_mapping

def import_csv(input_path, category, append_to_existing=True):
    if not os.path.exists(input_path):
        print(f"Error: Input CSV file '{input_path}' not found.")
        sys.exit(1)
        
    db_file_map = {
        "facilities": "../src/db/uganda_facilities.csv",
        "inventory": "../src/db/uganda_inventory.csv",
        "diseases": "../src/db/uganda_diseases.csv"
    }
    
    # Resolve correct paths
    script_dir = os.path.dirname(os.path.abspath(__file__))
    target_db_path = os.path.join(script_dir, db_file_map[category])
    
    print(f"Reading raw Kaggle/HDX dataset from: {input_path}")
    
    with open(input_path, mode='r', encoding='utf-8-sig') as f:
        reader = csv.reader(f)
        try:
            headers = next(reader)
        except StopIteration:
            print("Error: Input CSV is empty.")
            sys.exit(1)
            
    auto_mapped = auto_detect_mapping(headers, category)
    mapping = prompt_manual_mapping(headers, category, auto_mapped)
    
    raw_records = []
    with open(input_path, mode='r', encoding='utf-8-sig') as f:
        reader = csv.DictReader(f)
        for row in reader:
            raw_records.append(row)
            
    new_records = []
    
    # Process and map records
    for i, raw in enumerate(raw_records):
        new_row = {}
        
        if category == "facilities":
            new_row["id"] = raw.get(mapping["id"]) if mapping.get("id") else (i + 1)
            new_row["name"] = clean_value(raw.get(mapping["name"])) if mapping.get("name") else f"Facility {i+1}"
            new_row["type"] = clean_value(raw.get(mapping["type"])) if mapping.get("type") else "Health Center"
            
            lat_val = raw.get(mapping["latitude"]) if mapping.get("latitude") else "0.0"
            lng_val = raw.get(mapping["longitude"]) if mapping.get("longitude") else "0.0"
            try:
                new_row["latitude"] = float(lat_val)
            except (ValueError, TypeError):
                new_row["latitude"] = 0.0
            try:
                new_row["longitude"] = float(lng_val)
            except (ValueError, TypeError):
                new_row["longitude"] = 0.0
                
            new_row["contact"] = clean_value(raw.get(mapping["contact"])) if mapping.get("contact") else ""
            
        elif category == "inventory":
            new_row["name"] = clean_value(raw.get(mapping["name"])) if mapping.get("name") else f"Medicine {i+1}"
            
            qty_val = raw.get(mapping["quantity"]) if mapping.get("quantity") else "0"
            thresh_val = raw.get(mapping["minimumThreshold"]) if mapping.get("minimumThreshold") else "10"
            try:
                new_row["quantity"] = int(float(qty_val))
            except (ValueError, TypeError):
                new_row["quantity"] = 0
            try:
                new_row["minimumThreshold"] = int(float(thresh_val))
            except (ValueError, TypeError):
                new_row["minimumThreshold"] = 10
                
            new_row["unit"] = clean_value(raw.get(mapping["unit"])) if mapping.get("unit") else "units"
            new_row["lastUpdated"] = clean_value(raw.get(mapping["lastUpdated"])) if mapping.get("lastUpdated") else datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%SZ')
            
        elif category == "diseases":
            new_row["id"] = raw.get(mapping["id"]) if mapping.get("id") else (i + 1)
            new_row["title"] = clean_value(raw.get(mapping["title"])) if mapping.get("title") else f"Disease {i+1}"
            new_row["title_lg"] = clean_value(raw.get(mapping["title_lg"])) if mapping.get("title_lg") else ""
            new_row["icon"] = clean_value(raw.get(mapping["icon"])) if mapping.get("icon") else "medical-bag"
            new_row["color"] = clean_value(raw.get(mapping["color"])) if mapping.get("color") else "#3182CE"
            
            # Safe JSON steps parsing
            steps_raw = raw.get(mapping["steps_json"]) if mapping.get("steps_json") else "[]"
            steps_lg_raw = raw.get(mapping["steps_lg_json"]) if mapping.get("steps_lg_json") else "[]"
            
            # If not already JSON, split by comma or semi-colon
            def parse_steps(s):
                s = clean_value(s)
                if not s:
                    return []
                if (s.startswith("[") and s.endswith("]")):
                    try:
                        return json.loads(s)
                    except json.JSONDecodeError:
                        pass
                return [item.strip() for item in s.split(";") if item.strip()]
                
            new_row["steps_json"] = json.dumps(parse_steps(steps_raw))
            new_row["steps_lg_json"] = json.dumps(parse_steps(steps_lg_raw))
            
        new_records.append(new_row)
        
    # Read existing DB records if appending
    existing_records = []
    existing_keys = set()
    
    if append_to_existing and os.path.exists(target_db_path):
        with open(target_db_path, mode='r', encoding='utf-8') as f:
            reader = csv.DictReader(f)
            for row in reader:
                existing_records.append(row)
                if category == "facilities":
                    existing_keys.add(row["name"].lower())
                elif category == "inventory":
                    existing_keys.add(row["name"].lower())
                elif category == "diseases":
                    existing_keys.add(row["title"].lower())
                    
    # Merge and avoid duplicates
    merged_records = list(existing_records)
    added_count = 0
    
    for rec in new_records:
        key = rec["name"].lower() if category in ["facilities", "inventory"] else rec["title"].lower()
        if key in existing_keys:
            # Overwrite/update existing
            for idx, ext in enumerate(merged_records):
                ext_key = ext["name"].lower() if category in ["facilities", "inventory"] else ext["title"].lower()
                if ext_key == key:
                    merged_records[idx] = rec
                    break
        else:
            merged_records.append(rec)
            existing_keys.add(key)
            added_count += 1
            
    # Sort or fix IDs
    if category in ["facilities", "diseases"]:
        for idx, rec in enumerate(merged_records):
            rec["id"] = idx + 1
            
    # Write back to target DB path
    target_schema = SCHEMA_FACILITIES if category == "facilities" else (SCHEMA_INVENTORY if category == "inventory" else SCHEMA_DISEASES)
    
    print(f"\nWriting to target seed file: {target_db_path}")
    with open(target_db_path, mode='w', encoding='utf-8', newline='') as f:
        writer = csv.DictWriter(f, fieldnames=target_schema)
        writer.writeheader()
        for rec in merged_records:
            # Filter dict keys to standard schema
            filtered_rec = {k: rec.get(k, "") for k in target_schema}
            writer.writerow(filtered_rec)
            
    print(f"Success! Imported {len(new_records)} records from Kaggle/HDX.")
    print(f"Total records in system database seed: {len(merged_records)} (Added {added_count} new, updated {len(new_records) - added_count}).")
    
    # Prompt to update TypeScript seeds
    update_ts = input("\nWould you like to automatically rebuild the Typescript Seeds (csvSeeds.ts)? [Y/n]: ").strip().lower()
    if update_ts in ["", "y", "yes"]:
        rebuild_ts_seeds()

def rebuild_ts_seeds():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    db_dir = os.path.join(script_dir, "../src/db")
    
    fac_path = os.path.join(db_dir, "uganda_facilities.csv")
    inv_path = os.path.join(db_dir, "uganda_inventory.csv")
    dis_path = os.path.join(db_dir, "uganda_diseases.csv")
    seeds_ts_path = os.path.join(db_dir, "csvSeeds.ts")
    
    print(f"Rebuilding {seeds_ts_path}...")
    
    def read_csv_str(p):
        if os.path.exists(p):
            with open(p, 'r', encoding='utf-8') as f:
                return f.read().strip()
        return ""
        
    fac_str = read_csv_str(fac_path)
    inv_str = read_csv_str(inv_path)
    dis_str = read_csv_str(dis_path)
    
    ts_content = f"""export const FACILITIES_CSV = `{fac_str}`;

export const INVENTORY_CSV = `{inv_str}`;

export const DISEASES_CSV = `{dis_str}`;
"""
    
    with open(seeds_ts_path, 'w', encoding='utf-8') as f:
        f.write(ts_content)
    print("TypeScript Seeds rebuilt successfully!")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Kaggle/HDX CSV Dynamic Schema Importer for HealthGuard Uganda")
    parser.add_argument("input_csv", help="Path to raw Kaggle/HDX downloaded CSV file")
    parser.add_argument("category", choices=["facilities", "inventory", "diseases"], help="Data category to import")
    parser.add_argument("--no-append", action="store_true", help="Overwrite existing database seed instead of merging")
    parser.add_argument("--rebuild-only", action="store_true", help="Skip import, just rebuild Typescript seeds from current CSV files")
    
    args = parser.parse_args()
    
    if args.rebuild_only:
        rebuild_ts_seeds()
    else:
        import_csv(args.input_csv, args.category, append_to_existing=not args.no_append)

File: ml/import_kaggle_dataset.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');
const readline = require('readline');

const SCHEMA_FACILITIES = ["id", "name", "type", "latitude", "longitude", "contact"];
const SCHEMA_INVENTORY = ["name", "quantity", "unit", "minimumThreshold", "lastUpdated"];
const SCHEMA_DISEASES = ["id", "title", "title_lg", "icon", "color", "steps_json", "steps_lg_json"];

const MAPPINGS = {
    facilities: {
        name: ["facility", "clinic", "hospital", "name", "facility_name", "hosp_name", "title"],
        type: ["type", "level", "tier", "facility_type", "class", "category"],
        latitude: ["latitude", "lat", "y", "coord_lat", "latitude_deg"],
        longitude: ["longitude", "long", "lon", "x", "coord_long", "longitude_deg"],
        contact: ["contact", "phone", "telephone", "mobile", "email", "address"]
    },
    inventory: {
        name: ["name", "medicine", "drug", "item", "product", "description"],
        quantity: ["quantity", "qty", "stock", "count", "amount"],
        unit: ["unit", "pack", "form", "dosage_form"],
        minimumThreshold: ["minimumthreshold", "min_stock", "threshold", "min_qty", "alert_level"],
        lastUpdated: ["lastupdated", "updated", "date", "timestamp"]
    },
    diseases: {
        id: ["id", "code", "disease_id"],
        title: ["title", "name", "disease", "ailment", "condition"],
        title_lg: ["title_lg", "luganda", "local_name", "title_local"],
        icon: ["icon", "symbol", "image"],
        color: ["color", "hex", "theme"],
        steps_json: ["steps_json", "steps", "treatment", "guidelines", "protocol"],
        steps_lg_json: ["steps_lg_json", "steps_lg", "treatment_lg", "guidelines_lg"]
    }
};

// RFC-compliant CSV parser in JavaScript
function parseCSV(csvText) {
    const result = [];
    let row = [];
    let currentField = '';
    let inQuotes = false;

    for (let i = 0; i < csvText.length; i++) {
        const char = csvText[i];
        const nextChar = csvText[i + 1];

        if (inQuotes) {
            if (char === '"') {
                if (nextChar === '"') {
                    currentField += '"';
                    i++;
                } else {
                    inQuotes = false;
                }
            } else {
                currentField += char;
            }
        } else {
            if (char === '"') {
                inQuotes = true;
            } else if (char === ',') {
                row.push(currentField.trim());
                currentField = '';
            } else if (char === '\r' || char === '\n') {
                row.push(currentField.trim());
                if (row.length > 1 || row[0] !== '') {
                    result.push(row);
                }
                row = [];
                currentField = '';
                if (char === '\r' && nextChar === '\n') {
                    i++;
                }
            } else {
                currentField += char;
            }
        }
    }

    if (currentField !== '' || row.length > 0) {
        row.push(currentField.trim());
        if (row.length > 1 || row[0] !== '') {
            result.push(row);
        }
    }

    return result;
}

// Convert objects to CSV safely escaping quotes
function toCSV(records, headers) {
    const lines = [headers.join(',')];
    for (const rec of records) {
        const row = headers.map(h => {
            let val = rec[h] === undefined || rec[h] === null ? '' : String(rec[h]);
            if (val.includes(',') || val.includes('"') || val.includes('\n') || val.includes('\r')) {
                val = '"' + val.replace(/"/g, '""') + '"';
            }
            return val;
        });
        lines.push(row.join(','));
    }
    return lines.join('\n');
}

function autoDetectMapping(headers, category) {
    const detected = {};
    const categoryMappings = MAPPINGS[category];
    
    for (const [targetCol, synonyms] of Object.entries(categoryMappings)) {
        let found = false;
        for (const syn of synonyms) {
            for (const h of headers) {
                if (syn.toLowerCase() === h.trim().toLowerCase() || h.trim().toLowerCase().replace(/\s+/g, '_').includes(syn.toLowerCase())) {
                    detected[targetCol] = h;
                    found = true;
                    break;
                }
            }
            if (found) break;
        }
    }
    return detected;
}

function askQuestion(query) {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
    });
    return new Promise(resolve => rl.question(query, ans => {
        rl.close();
        resolve(ans.trim());
    }));
}

async function promptManualMapping(headers, category, autoMapped) {
    console.log(`\n--- Column Mapping Configuration for [${category.toUpperCase()}] ---`);
    console.log("Detected columns in raw CSV:");
    headers.forEach((h, idx) => {
        console.log(`  [${idx}] ${h}`);
    });
    console.log("\nWe need to map these to our database schema. Press Enter to accept auto-detection.");
    
    const finalMapping = {};
    const targetSchema = category === "facilities" ? SCHEMA_FACILITIES : (category === "inventory" ? SCHEMA_INVENTORY : SCHEMA_DISEASES);
    
    for (const target of targetSchema) {
        const defaultVal = autoMapped[target] || "";
        const defaultPrompt = defaultVal ? ` (Auto-detected: '${defaultVal}')` : "";
        
        while (true) {
            const choice = await askQuestion(`Select source column for '${target}'${defaultPrompt} [Enter Index/Name, or 'skip']: `);
            
            if (choice === "") {
                if (defaultVal) {
                    finalMapping[target] = defaultVal;
                    break;
                } else if (["contact", "title_lg", "steps_lg_json", "id", "icon", "color"].includes(target)) {
                    console.log(`Skipping optional field '${target}'`);
                    finalMapping[target] = null;
                    break;
                } else {
                    console.log(`Error: Field '${target}' is required. Please map it.`);
                }
            } else if (choice.toLowerCase() === "skip") {
                finalMapping[target] = null;
                break;
            } else {
                // Check if index
                const idx = parseInt(choice, 10);
                if (!isNaN(idx) && idx >= 0 && idx < headers.length) {
                    finalMapping[target] = headers[idx];
                    break;
                } else {
                    // Check if exact header match
                    const match = headers.find(h => h.toLowerCase() === choice.toLowerCase());
                    if (match) {
                        finalMapping[target] = match;
                        break;
                    } else {
                        console.log(`Header '${choice}' not found in raw CSV. Try again.`);
                    }
                }
            }
        }
    }
    return finalMapping;
}

function rebuildTsSeeds() {
    const dbDir = path.join(__dirname, "../src/db");
    const facPath = path.join(dbDir, "uganda_facilities.csv");
    const invPath = path.join(dbDir, "uganda_inventory.csv");
    const disPath = path.join(dbDir, "uganda_diseases.csv");
    const seedsTsPath = path.join(dbDir, "csvSeeds.ts");
    
    console.log(`Rebuilding ${seedsTsPath}...`);
    
    const readCSVStr = (p) => {
        if (fs.existsSync(p)) {
            return fs.readFileSync(p, 'utf8').trim();
        }
        return "";
    };
    
    const facStr = readCSVStr(facPath);
    const invStr = readCSVStr(invPath);
    const disStr = readCSVStr(disPath);
    
    const tsContent = `export const FACILITIES_CSV = \`${facStr}\`;

export const INVENTORY_CSV = \`${invStr}\`;

export const DISEASES_CSV = \`${disStr}\`;
`;
    
    fs.writeFileSync(seedsTsPath, tsContent, 'utf8');
    console.log("TypeScript Seeds rebuilt successfully!");
}

async function run() {
    const args = process.argv.slice(2);
    
    if (args.includes("--rebuild-only")) {
        rebuildTsSeeds();
        return;
    }
    
    if (args.length < 2) {
        console.log("Usage: node import_kaggle_dataset.js <input_csv_path> <category: facilities|inventory|diseases> [--no-append]");
        console.log("Or to rebuild ts seeds: node import_kaggle_dataset.js --rebuild-only");
        return;
    }
    
    const inputPath = args[0];
    const category = args[1];
    const noAppend = args.includes("--no-append");
    const appendToExisting = !noAppend;
    
    if (!["facilities", "inventory", "diseases"].includes(category)) {
        console.log("Error: Category must be one of: facilities, inventory, diseases");
        return;
    }
    
    if (!fs.existsSync(inputPath)) {
        console.log(`Error: Input CSV file '${inputPath}' not found.`);
        return;
    }
    
    const dbFileMap = {
        facilities: "../src/db/uganda_facilities.csv",
        inventory: "../src/db/uganda_inventory.csv",
        diseases: "../src/db/uganda_diseases.csv"
    };
    
    const targetDbPath = path.join(__dirname, dbFileMap[category]);
    
    console.log(`Reading raw Kaggle/HDX dataset from: ${inputPath}`);
    const csvContent = fs.readFileSync(inputPath, 'utf8');
    const parsedRows = parseCSV(csvContent);
    
    if (parsedRows.length <= 1) {
        console.log("Error: Input CSV is empty or only contains headers.");
        return;
    }
    
    const headers = parsedRows[0];
    const autoMapped = autoDetectMapping(headers, category);
    const mapping = await promptManualMapping(headers, category, autoMapped);
    
    const rawRecords = [];
    for (let i = 1; i < parsedRows.length; i++) {
        const row = parsedRows[i];
        const record = {};
        headers.forEach((h, idx) => {
            record[h] = row[idx] || "";
        });
        rawRecords.append || rawRecords.push(record);
    }
    
    const newRecords = [];
    
    rawRecords.forEach((raw, i) => {
        const newRow = {};
        
        if (category === "facilities") {
            newRow["id"] = mapping["id"] ? raw[mapping["id"]] : (i + 1);
            newRow["name"] = mapping["name"] ? raw[mapping["name"]].trim() : `Facility ${i+1}`;
            newRow["type"] = mapping["type"] ? raw[mapping["type"]].trim() : "Health Center";
            
            const latVal = mapping["latitude"] ? raw[mapping["latitude"]] : "0.0";
            const lngVal = mapping["longitude"] ? raw[mapping["longitude"]] : "0.0";
            newRow["latitude"] = parseFloat(latVal) || 0.0;
            newRow["longitude"] = parseFloat(lngVal) || 0.0;
            newRow["contact"] = mapping["contact"] ? raw[mapping["contact"]].trim() : "";
            
        } else if (category === "inventory") {
            newRow["name"] = mapping["name"] ? raw[mapping["name"]].trim() : `Medicine ${i+1}`;
            
            const qtyVal = mapping["quantity"] ? raw[mapping["quantity"]] : "0";
            const threshVal = mapping["minimumThreshold"] ? raw[mapping["minimumThreshold"]] : "10";
            newRow["quantity"] = parseInt(qtyVal, 10) || 0;
            newRow["minimumThreshold"] = parseInt(threshVal, 10) || 10;
            newRow["unit"] = mapping["unit"] ? raw[mapping["unit"]].trim() : "units";
            newRow["lastUpdated"] = mapping["lastUpdated"] ? raw[mapping["lastUpdated"]].trim() : new Date().toISOString();
            
        } else if (category === "diseases") {
            newRow["id"] = mapping["id"] ? raw[mapping["id"]] : (i + 1);
            newRow["title"] = mapping["title"] ? raw[mapping["title"]].trim() : `Disease ${i+1}`;
            newRow["title_lg"] = mapping["title_lg"] ? raw[mapping["title_lg"]].trim() : "";
            newRow["icon"] = mapping["icon"] ? raw[mapping["icon"]].trim() : "medical-bag";
            newRow["color"] = mapping["color"] ? raw[mapping["color"]].trim() : "#3182CE";
            
            const stepsRaw = mapping["steps_json"] ? raw[mapping["steps_json"]] : "[]";
            const stepsLgRaw = mapping["steps_lg_json"] ? raw[mapping["steps_lg_json"]] : "[]";
            
            const parseSteps = (s) => {
                s = (s || "").trim();
                if (!s) return [];
                if (s.startsWith("[") && s.endsWith("]")) {
                    try { return JSON.parse(s); } catch (e) {}
                }
                return s.split(";").map(item => item.trim()).filter(Boolean);
            };
            
            newRow["steps_json"] = JSON.stringify(parseSteps(stepsRaw));
            newRow["steps_lg_json"] = JSON.stringify(parseSteps(stepsLgRaw));
        }
        
        newRecords.push(newRow);
    });
    
    // Read existing DB records
    let existingRecords = [];
    const existingKeys = new Set();
    
    if (appendToExisting && fs.existsSync(targetDbPath)) {
        const dbContent = fs.readFileSync(targetDbPath, 'utf8');
        const dbRows = parseCSV(dbContent);
        if (dbRows.length > 1) {
            const dbHeaders = dbRows[0];
            for (let i = 1; i < dbRows.length; i++) {
                const row = dbRows[i];
                const rec = {};
                dbHeaders.forEach((h, idx) => {
                    rec[h] = row[idx] || "";
                });
                existingRecords.push(rec);
                
                const key = category === "facilities" ? rec["name"].toLowerCase() :
                            category === "inventory" ? rec["name"].toLowerCase() :
                            rec["title"].toLowerCase();
                existingKeys.add(key);
            }
        }
    }
    
    const mergedRecords = [...existingRecords];
    let addedCount = 0;
    
    newRecords.forEach(rec => {
        const key = category === "facilities" ? rec["name"].toLowerCase() :
                    category === "inventory" ? rec["name"].toLowerCase() :
                    rec["title"].toLowerCase();
                    
        if (existingKeys.has(key)) {
            // Overwrite/update existing
            const idx = mergedRecords.findIndex(ext => {
                const extKey = category === "facilities" ? ext["name"].toLowerCase() :
                               category === "inventory" ? ext["name"].toLowerCase() :
                               ext["title"].toLowerCase();
                return extKey === key;
            });
            if (idx !== -1) {
                mergedRecords[idx] = rec;
            }
        } else {
            mergedRecords.push(rec);
            existingKeys.add(key);
            addedCount++;
        }
    });
    
    // Sort or fix IDs
    if (["facilities", "diseases"].includes(category)) {
        mergedRecords.forEach((rec, idx) => {
            rec["id"] = idx + 1;
        });
    }
    
    const targetSchema = category === "facilities" ? SCHEMA_FACILITIES : (category === "inventory" ? SCHEMA_INVENTORY : SCHEMA_DISEASES);
    const outputCSV = toCSV(mergedRecords, targetSchema);
    
    console.log(`\nWriting to target seed file: ${targetDbPath}`);
    fs.writeFileSync(targetDbPath, outputCSV, 'utf8');
    
    console.log(`Success! Imported ${newRecords.length} records.`);
    console.log(`Total records in system seed: ${mergedRecords.length} (Added ${addedCount} new, updated ${newRecords.length - addedCount}).`);
    
    const ans = await askQuestion("\nWould you like to automatically rebuild the Typescript Seeds (csvSeeds.ts)? [Y/n]: ");
    if (ans.toLowerCase() === "" || ans.toLowerCase() === "y" || ans.toLowerCase() === "yes") {
        rebuildTsSeeds();
    }
}

if (require.main === module) {
    run().catch(err => {
        console.error("An error occurred:", err);
    });
}
